src/values/callable_value.rs

```diff
@@ -0,0 +1,127 @@
+use std::convert::{TryFrom};
+use either::Either;
+
+use crate::values::{AsValueRef, BasicValueEnum, InstructionValue, Value};
+use crate::values::FunctionValue;
+use crate::values::PointerValue;
+
+use llvm_sys::prelude::LLVMValueRef;
+use llvm_sys::core::{LLVMGetTypeKind, LLVMGetElementType, LLVMTypeOf, LLVMGetReturnType};
+use llvm_sys::LLVMTypeKind;
+
+use crate::builder::Builder;
+
+/// A value that can be called with the [`Builder::build_call`] instruction.
+///
+/// In practice, the `F : Into<CallableValue<'ctx>>` bound of [`Builder::build_call`] means it is
+/// possible to pass a [`FunctionValue`] to `build_call` directly. It will be implicitly converted
+/// into a `CallableValue`.
+///
+/// ```no_run
+/// use inkwell::context::Context;
+///
+/// // A simple function which calls itself:
+/// let context = Context::create();
+/// let module = context.create_module("ret");
+/// let builder = context.create_builder();
+/// let i32_type = context.i32_type();
+/// let fn_type = i32_type.fn_type(&[i32_type.into()], false);
+/// let fn_value = module.add_function("ret", fn_type, None);
+/// let entry = context.append_basic_block(fn_value, "entry");
+/// let i32_arg = fn_value.get_first_param().unwrap();
+///
+/// builder.position_at_end(entry);
+///
+/// let ret_val = builder.build_call(fn_value, &[i32_arg], "call")
+///     .try_as_basic_value()
+///     .left()
+///     .unwrap();
+///
+/// builder.build_return(Some(&ret_val));
+/// ```
+///
+/// A [`PointerValue`] cannot be implicitly converted to a `CallableValue` because the pointer may
+/// point to a non-function value. Instead we can use [`TryFrom`] and [`TryInto`] to handle this
+/// failure case explicitly.
+///
+/// ```no_run
+/// use std::convert::TryFrom;
+/// use inkwell::context::Context;
+/// use inkwell::values::CallableValue;
+///
+/// // A simple function which calls itself:
+/// let context = Context::create();
+/// let module = context.create_module("ret");
+/// let builder = context.create_builder();
+/// let i32_type = context.i32_type();
+/// let fn_type = i32_type.fn_type(&[i32_type.into()], false);
+/// let fn_value = module.add_function("ret", fn_type, None);
+/// let entry = context.append_basic_block(fn_value, "entry");
+/// let i32_arg = fn_value.get_first_param().unwrap();
+///
+/// builder.position_at_end(entry);
+///
+/// // take a pointer to the function value
+/// let fn_pointer_value = fn_value.as_global_value().as_pointer_value();
+///
+/// // convert that pointer value into a callable value
+/// // explicitly handling the failure case (here with `unwrap`)
+/// let callable_value = CallableValue::try_from(fn_pointer_value).unwrap();
+///
+/// let ret_val = builder.build_call(callable_value, &[i32_arg], "call")
+///     .try_as_basic_value()
+///     .left()
+///     .unwrap();
+///
+/// builder.build_return(Some(&ret_val));
+/// ```
+#[derive(Debug)]
+pub struct CallableValue<'ctx>(FunctionOrPointerValue<'ctx>);
```

Comment:
Is `FunctionOrPointerValue` used anywhere else? It kinda seems like `CallableValue` should just take its place

Reply:
Nvm, I see the type alias below